programs/fileio.c

```diff
@@ -812,41 +836,134 @@ FIO_compressZstdFrame(const cRess_t* ressPtr,
+            if (!toFlushNow) flushWaiting = 1;
 
             /* Write compressed stream */
-            DISPLAYLEVEL(6, "ZSTD_compress_generic(end:%u) => intput pos(%u)<=(%u)size ; output generated %u bytes \n",
+            DISPLAYLEVEL(6, "ZSTD_compress_generic(end:%u) => input pos(%u)<=(%u)size ; output generated %u bytes \n",
                             (U32)directive, (U32)inBuff.pos, (U32)inBuff.size, (U32)outBuff.pos);
             if (outBuff.pos) {
                 size_t const sizeCheck = fwrite(ress.dstBuffer, 1, outBuff.pos, dstFile);
-                if (sizeCheck!=outBuff.pos)
+                if (sizeCheck != outBuff.pos)
                     EXM_THROW(25, "Write error : cannot write compressed block");
                 compressedfilesize += outBuff.pos;
             }
+
+            /* display notification; and adapt compression level */
             if (READY_FOR_UPDATE()) {
                 ZSTD_frameProgression const zfp = ZSTD_getFrameProgression(ress.cctx);
                 double const cShare = (double)zfp.produced / (zfp.consumed + !zfp.consumed/*avoid div0*/) * 100;
+
+                /* display progress notifications */
                 if (g_displayLevel >= 3) {
-                    DISPLAYUPDATE(3, "\r(L%i) Buffered :%4u MB - Consumed :%4u MB - Compressed :%4u MB => %.2f%%",
+                    DISPLAYUPDATE(3, "\r(L%i) Buffered :%4u MB - Consumed :%4u MB - Compressed :%4u MB => %.2f%% ",
                                 compressionLevel,
                                 (U32)((zfp.ingested - zfp.consumed) >> 20),
                                 (U32)(zfp.consumed >> 20),
                                 (U32)(zfp.produced >> 20),
                                 cShare );
-                } else {
-                    /* g_displayLevel <= 2; only display notifications if == 2; */
+                } else {   /* summarized notifications if == 2; */
                     DISPLAYLEVEL(2, "\rRead : %u ", (U32)(zfp.consumed >> 20));
                     if (fileSize != UTIL_FILESIZE_UNKNOWN)
                         DISPLAYLEVEL(2, "/ %u ", (U32)(fileSize >> 20));
                     DISPLAYLEVEL(2, "MB ==> %2.f%% ", cShare);
                     DELAY_NEXT_UPDATE();
                 }
-            }
-        }
+
+                /* adaptive mode : statistics measurement and speed correction */
+                if (g_adaptiveMode) {
+
+                    /* check output speed */
+                    if (zfp.currentJobID > 1) {  /* only possible if nbWorkers >= 1 */
+
+                        unsigned long long newlyProduced = zfp.produced - previous_zfp_update.produced;
+                        unsigned long long newlyFlushed = zfp.flushed - previous_zfp_update.flushed;
+                        assert(zfp.produced >= previous_zfp_update.produced);
+                        assert(g_nbWorkers >= 1);
+
+                        if ( (zfp.ingested == previous_zfp_update.ingested)   /* no data read : input buffer full */
```

Comment:
Doesn't this condition suggest that compression is too slow? When would this happen if compression is too fast?

Reply:
In the general case, yes : if ingestion is blocked, it means input is faster than compression can consume.

But this case is a bit different : we are actually testing a situation where both compression and input are blocked because output speed is too slow.

In which case, it doesn't matter that input is faster : compression should nonetheless slow down in an attempt to match output speed.

I shall add a comment to explain the nature of this test.
Also, thinking about it again, it seems I don't really need to check ingestion rate. The test should work too with just consumption rate and `nbActiveWorkers` (which is a recently added indicator).
I will make more tests to check that hypothesis.

Reply:
I kind of remember now that controlling ingestion rate was meant to cover the initial part,
when the compressor is waiting for at least one job to be filled,
and cannot even start as long as the first buffer is not completely filled.

In such a case, `nbActiveWorkers = 0` and `consumptionRate == 0`. But it doesn't mean that output speed is too slow. It's actually the reverse situation : input speed is too slow.

Note that the outcome is still the same : compression must slow down.

But input speed is too slow _compared to what_ ?
While waiting for the first job, it's only compared to the notification update rate. 
We don't know yet if input speed is too slow compared to compression speed.
If initial compression level is set to 19, it might very well be fast enough.
It's just we did not had a chance yet to compare both.

So, in order to avoid triggering a "slow down" order while waiting for the first job to fill up, I added a test looking at ingestion rate.
Since then, test order has been changed, and now this test is behind an `if (zfp.currentJobID > 1)` condition, which means the first job has long been started.

Now, a situation where all buffers are empty, resulting in compression waiting for more input, may still happen in the future : we could imagine a sudden prolonged drop in input speed. In which case, measuring `consumptionRate` and `nbActiveWorkers` will trigger this condition.
But this time, this is really because input is too slow. So we can slow down compression.

Hence this test can be simplified by removing the `ingestionRate` part.